File: web/src/main/java/top/bettercode/summer/web/deprecated/DeprecatedAPIInterceptor.java
package top.bettercode.summer.web.deprecated;

import java.util.Locale;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.springframework.context.MessageSource;
import org.springframework.web.method.HandlerMethod;
import top.bettercode.summer.tools.lang.util.AnnotatedUtils;
import top.bettercode.summer.web.servlet.NotErrorHandlerInterceptor;

/**
 * 已弃用的接口检查
 *
 * @author Peter Wu
 */
public class DeprecatedAPIInterceptor implements NotErrorHandlerInterceptor {

  private final MessageSource messageSource;

  public DeprecatedAPIInterceptor(MessageSource messageSource) {
    this.messageSource = messageSource;
  }

  @Override
  public boolean preHandlerMethod(HttpServletRequest request, HttpServletResponse response,
      HandlerMethod handler) {
    DeprecatedAPI annotation = AnnotatedUtils.getAnnotation(handler,
        DeprecatedAPI.class);
    if (annotation != null) {
      throw new IllegalStateException(getText(request, annotation.message()));
    }
    return true;
  }

  private String getText(HttpServletRequest request, Object code, Object... args) {
    String codeString = String.valueOf(code);
    return messageSource.getMessage(codeString, args, codeString,
        request == null ? Locale.CHINA : request.getLocale());
  }

}

File: web/src/main/kotlin/top/bettercode/summer/web/deprecated/DeprecatedAPIInterceptor.kt
package top.bettercode.summer.web.deprecated

import org.springframework.context.MessageSource
import org.springframework.web.method.HandlerMethod
import top.bettercode.summer.tools.lang.util.AnnotatedUtils.getAnnotation
import top.bettercode.summer.web.servlet.NotErrorHandlerInterceptor
import java.util.*
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * 已弃用的接口检查
 *
 * @author Peter Wu
 */
class DeprecatedAPIInterceptor(private val messageSource: MessageSource) : NotErrorHandlerInterceptor {
    override fun preHandlerMethod(request: HttpServletRequest?, response: HttpServletResponse?,
                                  handler: HandlerMethod?): Boolean {
        val annotation = getAnnotation(handler!!,
                DeprecatedAPI::class.java)
        check(annotation == null) { getText(request, annotation!!.message) }
        return true
    }

    private fun getText(request: HttpServletRequest?, code: Any, vararg args: Any): String {
        val codeString = code.toString()
        return messageSource.getMessage(codeString, args, codeString,
                if (request == null) Locale.CHINA else request.locale)?:""
    }
}
